core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/tri/BavetJoinTriConstraintStream.java

```diff
@@ -101,7 +97,29 @@ public ConstraintStream getTupleSource() {
     // Equality for node sharing
     // ************************************************************************
 
-    // TODO
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        BavetJoinTriConstraintStream<?, ?, ?, ?> other = (BavetJoinTriConstraintStream<?, ?, ?, ?>) o;
+        /*
+         * Bridge streams do not implement equality because their equals() would have to point back to this stream,
+         * resulting in StackOverflowError.
+         * Therefore we need to check bridge parents to see where this join node comes from.
```

Comment:
Aaaaah. That explains. Let's talk :)

Reply:
I suspect the need for the active set to be an IdentityHashSet to be a side-effect of this exception.
That is very wierd, because after the "stream sharing" all traces of such extra bridge nodes should be gone. Let's debug together.

Reply:
Besides this issue, everything else if fine to merge. I 'd like to understand the exception to the design first - see if we can avoid it or need to just accept+document it - before merging.